src/lib/index/index_read.c

```diff
@@ -241,6 +252,22 @@ int PUBLIC zck_get_chunk_valid(zckChunk *idx) {
     return idx->valid;
 }
 
+bool PUBLIC zck_set_chunk_valid(zckChunk *idx, int valid) {
```

Comment:
I don't like the idea of overloading idx->valid with a source id, and I don't see any other reason to be able to manually set a chunk as valid, so I'm going to NAK this function.  One thing that occured to me is that each chunk has the `zckCtx` that it's attached to, so running `zck_get_src_chunk` against each chunk in the target would get you the source chunk.  We could then create a function, `zck_get_chunk_context`, that would return the `zckCtx` for the source chunk.  Would that work?

Reply:
Ok, I take your point about needing alternate (non-http) ways of getting the chunks.  My concern is that we're setting the chunk's "valid" status without verifying that it is, indeed, valid.  I would prefer something that would work like `zck_validate_checksums`, but just with the individual chunk.  It would check the chunk against the checksum, set the valid tag to true, and return 1.  If the chunk is invalid, it would return -1, and if an error, it would return 0 as the other validation functions do.  Would something like this work for your use case?

Reply:
The point is that, if libzck is not responsible to put all chunks together to create the new file, is also not responsible for validating the chunks. Of course, this must be done and who is in charge to decide which chunks are valid must do it - no discussion about it ! However, a zck_validate_checksum() (even on single chunk) will require that the whole chunk should be stored (the current functions requires a seekable file), and this is a problem on system with low resources. It is pretty better to compute the hash "on the fly during" the download, without any temporary copy. The reason for me to implement zck_set_chunk_valid() is just to reuse zck_get_missing_range() as it is, because the range is built using the flag, but I can create the range on my own. Then I do not need the zckRange and the setter. 

As this is the last open thread in the PR, do you agree to merge Patches 1-6 by dropping "Add setter for chunk valid field " ? Do you need that I repush the branch without last commit, or you can simply merge it without it ?

Reply:
I am happy to merge patches 1-6 and just drop this one.  I'll go ahead and do that myself.

I do see your point about computing the hash on the fly during the download, but I'm still very leery of adding this function as-is, since it's a great way of shooting yourself in the foot.  Maybe we should do a voice/video call at some point to see if there's some way that we can assuage my concerns.  Maybe we can setup a callback that can be used with whatever downloader that would automatically hash and validate the chunks as they're downloaded (similar to how we're doing in libcurl)?

Reply:
I've now merged all except the last patch

Reply:
> I am happy to merge patches 1-6 and just drop this one. I'll go ahead and do that myself.
> 

Thanks !

> I do see your point about computing the hash on the fly during the download, but I'm still very leery of adding this function as-is, since it's a great way of shooting yourself in the foot. Maybe we should do a voice/video call at some point to see if there's some way that we can assuage my concerns.

Of course,  it will be nice to have it !

> Maybe we can setup a callback that can be used with whatever downloader that would automatically hash and validate the chunks as they're downloaded (similar to how we're doing in libcurl)?

Do not know - it maybe depends on use cases.